### PR TITLE
fetched products and categories and attributes and mapped them

### PR DESCRIPTION
**Pull request by Tejas-Choudekar**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -40,6 +40,8 @@ dependencies {
 	compile "com.sun.xml.bind:jaxb-core:2.3.0"
 	compile "com.sun.xml.ws:rt:2.3.0"
 	compile "com.sun.xml.ws:jaxws-rt:2.3.0"
+	compile "org.springframework.boot:spring-boot-starter-validation"
+	compile 'com.squareup.okhttp3:logging-interceptor:3.8.0'
     compile("org.springframework.boot:spring-boot-starter-data-jpa")
```
Comment (arohablue) [inline build.gradle:45]: Do not push   *.class files

File: build.gradle [hunk 2]
```diff
@@ -54,6 +56,19 @@ dependencies {
...
+
+	// https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient
+	compile group: 'org.apache.httpcomponents', name: 'httpclient', version: '4.5.11'
+
+
+	// https://mvnrepository.com/artifact/com.googlecode.json-simple/json-simple
+	compile group: 'com.googlecode.json-simple', name: 'json-simple', version: '1.1.1'
+
+	// https://mvnrepository.com/artifact/com.icoderman/wc-api-java
+	compile group: 'com.icoderman', name: 'wc-api-java', version: '1.4'
```
Comment (arohablue) [inline build.gradle:71]: keep proper spacing

File: src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java [hunk 3]
```diff
@@ -51,11 +76,58 @@ public ProductCategory getCategoryById(@RequestBody ProductCategory category){
...
+    }
+
     @PostMapping("/getcatgories")
     public List<ProductCategory> getcategories(){
         return productService.getCategories();
     }
 
+    @GetMapping("/layinproductsfromwoo")
+    public ResponseStatus layInProductsFromWoo() {
+        System.out.println("came here");
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java:91]: remove sout

File: src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java [hunk 4]
```diff
@@ -16,6 +34,13 @@
     @Autowired
     ProductService productService;
 
+    @Autowired
+    WoocommerceRequestService wooReq;
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java:38]: user proper naming it should be : woocommerceRequestService

File: src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java [hunk 5]
```diff
@@ -31,20 +37,25 @@
   @Column(name = "name")
   private String name; // Category Name
 
+  @Column(name = "wooId")
+  private Integer wooId;
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java:42]: user proper column name. its should be `woo_id`

File: src/main/java/com/example/resellerapp/resellerapp/repositories/CustomerOrderRepository.java [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 import org.springframework.data.jpa.repository.JpaRepository;
 
+//Adding Comments
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/repositories/CustomerOrderRepository.java:7]: why every file has "Adding Comments" comment on top?

File: src/main/java/com/example/resellerapp/resellerapp/repositories/ProductAttributeRepository.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.data.repository.query.Param;
+import java.util.List;
+import java.util.Set;
+
+public interface ProductAttributeRepository extends JpaRepository<ProductAttribute, String> {
+
+    Set<ProductAttribute> findByWooId(@Param("wooId") Integer wooId);
+    // List<ProductCategory>findByName(@Param("name") String name);
+    // ProductCategory findByWooId(@Param("wooId") Integer wooId);
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/repositories/ProductAttributeRepository.java:17]: do not comment

File: src/main/java/com/example/resellerapp/resellerapp/repositories/ProductDimensionRepository.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.List;
+import java.util.Set;
+
+public interface ProductDimensionRepository extends JpaRepository<ProductDimension, String> {
+
+    // ProductCategory findById(@Param("id") Long id);
+
+    // List<ProductCategory>findByName(@Param("name") String name);
+    // ProductCategory findByWooId(@Param("wooId") Integer wooId);
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/repositories/ProductDimensionRepository.java:19]: remove comments and add  proper methods for ProductDimension

File: src/main/java/com/example/resellerapp/resellerapp/repositories/ProductImageRepository.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import java.util.List;
+import java.util.Set;
+
+public interface ProductImageRepository extends JpaRepository<ProductImage, String> {
+
+    // ProductCategory findById(@Param("id") Long id);
+
+    // List<ProductCategory>findByName(@Param("name") String name);
+    // ProductCategory findByWooId(@Param("wooId") Integer wooId);
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/repositories/ProductImageRepository.java:19]: same as above. Add wherever necessary.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+    private ProductCategoryRepository productCategoryRepository;
+
+    @Autowired
+    private ProductAttributeRepository productAttributeRepository;
+
+    @Autowired
+    private ProductRepository productRepository;
+
+    OAuthConfig config = new OAuthConfig("https://www.sustainida.com", "ck_13a0ff1577f75a8c2d5943767e9cd60a3a27a93b",
+            "cs_7547a62231d9736e63c5ec76cedac9fafcb0eb77");
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:44]: Keep this separate in constants file for now. Each parameter stored separately.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+        // TODO Auto-generated method stub
+        return null;
+    }
+
+    @Override
+    public Boolean layInAllProductImages() {
+        // TODO Auto-generated method stub
+        return null;
+    }
+}
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:331]: Reflector this class into methods and will review after that,

File: src/main/java/com/util/Constants.java [hunk 12]
```diff
@@ -18,4 +19,10 @@
 	public static final int ADDRESS_ALREADY_PRESENT = 302;
 
 	public static final int USER_NOT_FOUND = 401;
+
+	public static final int CATEGORIES_ADDED_SUCCESSFULLY = 501;
+	public static final int CATEGORIES_ADDITION_FAILED = 502;
+	public static final int ATTRIBUTES_ADDED_SUCCESSFULLY = 503;
+	public static final int ATTRIBUTES_ADDITION_FAILED = 504;
+
```
Comment (arohablue) [inline src/main/java/com/util/Constants.java:27]: user proper codes. 
1. Success codes start with 1**  
2.failed code with 2**
3.Duplicate date with 3**
4.Not found with 4**

File: src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java [hunk 13]
```diff
@@ -1,63 +1,104 @@
...
     @Autowired
     ProductService productService;
 
-    @PostMapping("/getproducts")  //Returns all Products 
-    public List<Product> getProducts() {	
+    @Autowired
+    WoocommerceRequestService woocommerceRequestService;
+
+    Gson gson;
+    List<Object> product;
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/controllers/ProductController.java:38]: Where is this variable used: `List<Object> product;`

File: src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java (hunk 5 above)
Comment (arohablue) [reply src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java:42]: update this

File: src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java [hunk 14]
```diff
@@ -54,9 +65,10 @@ public int hashCode() {
     result = prime * result + ((description == null) ? 0 : description.hashCode());
     result = prime * result + ((id == null) ? 0 : id.hashCode());
     result = prime * result + ((name == null) ? 0 : name.hashCode());
-    result = prime * result + ((parentCategory == null) ? 0 : parentCategory.hashCode());
+    //result = prime * result + ((parentCategory == null) ? 0 : parentCategory.hashCode());
     result = prime * result + ((products == null) ? 0 : products.hashCode());
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java:69]: put comment why we did this

File: src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java [hunk 15]
```diff
@@ -126,22 +151,26 @@ public void setDescription(String description) {
...
     return products;
   }
 
   public void setProducts(Set<Product> products) {
     this.products = products;
   }
-  @JsonIgnore
+
+
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java:164]: why extra space ?

File: src/main/java/com/example/resellerapp/resellerapp/services/ProductService.java [hunk 16]
```diff
@@ -78,6 +79,17 @@ public ProductCategory getCategoryById(Long id){
         }
     }
 
+    @Override
+    public ProductCategory getCategoryByWooId(int id){
+        ProductCategory category = productCategoryRepository.findByWooId(id);
+        if(category != null){
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/ProductService.java:85]: what if id is null ?

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+import com.example.resellerapp.resellerapp.repositories.ProductImageRepository;
+import com.example.resellerapp.resellerapp.repositories.ProductRepository;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.icoderman.woocommerce.ApiVersionType;
+import com.icoderman.woocommerce.WooCommerce;
+import com.icoderman.woocommerce.WooCommerceAPI;
+import com.icoderman.woocommerce.oauth.OAuthConfig;
+import com.util.Constants;
+import com.icoderman.woocommerce.*;
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:36]: if you have used `woocommerce.*`
Then there is no need to use individual imports please only use one or the other

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+    SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");
+
+    OAuthConfig config = new OAuthConfig(Constants.URL, Constants.CONSUMER_KEY, Constants.CONSUMER_SECRET);
+
+    WooCommerce wooCommerce = new WooCommerceAPI(config, ApiVersionType.V1);
+
+    int i;
+    int j;
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:64]: remove these variables from here

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 19]
```diff
@@ -0,0 +1,438 @@
+package com.example.resellerapp.resellerapp.services;
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:1]: create a new helper class and move all the code which is being used again and again there.

Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:1]: Create a separate class for each of them and call them from here. Keep all data related calls here just pass them the Retrieved JSON.
1.Fetching Products
2.Fetching Categories

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 20]
```diff
@@ -0,0 +1,438 @@
...
+                    String storeThisProductRegularPrice = wooProducts.get(i).get("regular_price").toString();
+                    if (storeThisProductRegularPrice.equals(storeThisProductPrice)) {
+                        product.setRegularPrice(true);
+                    } else {
+                        product.setRegularPrice(false);
+                    }
+                } else {
+                    product.setRegularPrice(false);
+                }
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:143]: The above logic should be a separate method.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+                Boolean manageStock = (boolean) wooProducts.get(i).get("manage_stock");
+                product.setManageStock(manageStock);
+
+                if (wooProducts.get(i).get("stock_quantity") != null) {
+                    int storeStockQuantity = (int) wooProducts.get(i).get("stock_quantity");
+                    product.setStockQuantity(storeStockQuantity);
+                } else {
+                    product.setStockQuantity(0);
+                }
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:153]: Separate Method.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 22]
```diff
@@ -0,0 +1,438 @@
...
+                Boolean stockCheckerBoolean = (boolean) wooProducts.get(i).get("in_stock");
+                if (stockCheckerBoolean == true) {
+                    product.setStockStatus("In Stock");
+                } else {
+                    product.setStockStatus("Out Of Stock");
+                }
+
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:160]: Separate Method.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 23]
```diff
@@ -0,0 +1,438 @@
...
+                product.setWeight(storeThisProductWeight);
+
+                int parentId = (int) wooProducts.get(i).get("parent_id");
+                product.setParentId(parentId);
+
+                productRepository.save(product);
+
+            }
+        }
+        if (productRepository.count() >= 72) {
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:171]: Do not use hardcoded value. Retrieve count from the woo commerce and compare

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 24]
```diff
@@ -0,0 +1,438 @@
...
+                ProductCategory productCategory = productCategoryRepository.findByWooId(wooChildId);
+                ProductCategory categoryToSetAsParent = productCategoryRepository.findByWooId(wooParentId);
+
+                if (wooParentId != 0) {
+
+                    if (categoryToSetAsParent != null && productCategory != null) {
+
+                        productCategory.setParentCategory(categoryToSetAsParent);
+                        productCategoryRepository.save(productCategory);
+                    }
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:219]: seperate method

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 25]
```diff
@@ -0,0 +1,438 @@
...
+                }
+            }
+
+        }
+        if (productCategoryRepository.count() == 181) {
+            return true;
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:225]: same as above. Do not use hardcoded value.

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 26]
```diff
@@ -0,0 +1,438 @@
...
+
+    @Override
+    public Boolean mapProductAndAttributes(List<Map<String, Object>> wooProducts) {
+        Set<ProductAttribute> setOfProductAttribute;
+        for (i = 0; i < wooProducts.size(); i++) {
+            int storeThisAttributeId = (int) wooProducts.get(i).get("id");
+            setOfProductAttribute = new HashSet<ProductAttribute>();
+            setOfProductAttribute = productAttributeRepository.findByWooId(storeThisAttributeId);
+            if (!setOfProductAttribute.isEmpty() && setOfProductAttribute != null) {
+                System.out.println(setOfProductAttribute.size());
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:304]: no sout

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 27]
```diff
@@ -0,0 +1,438 @@
...
+            for (i = 0; i < wooProducts.size(); i++) {
+                if (!wooProducts.get(i).get("status").toString().equals("draft")) {
+                    List<Map<String, Object>> storeTheseImages = (List<Map<String, Object>>) wooProducts.get(i)
+                            .get("images");
+                    if (!storeTheseImages.isEmpty()) {
+                        ProductImage productImage = new ProductImage();
+                        Product product = new Product();
+                        int id = (int) wooProducts.get(i).get("id");
+                        for (j = 0; j < storeTheseImages.size(); j++) {
+                            System.out.println(id);
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:378]: remove System.out.println

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 28]
```diff
@@ -0,0 +1,438 @@
...
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public Boolean layInAllProductAttributes(List<Map<String, Object>> wooProducts) {
+
+        if (wooProducts != null) {
+            System.out.println(wooProducts.size());
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:259]: remove System.out.println

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java [hunk 29]
```diff
@@ -0,0 +1,438 @@
...
+                                wooDateCreated = formatter
+                                        .parse(storeTheseImages.get(j).get("date_created").toString());
+                                productImage.setDateCreated(wooDateCreated);
+
+                                wooDateModified = formatter
+                                        .parse(storeTheseImages.get(j).get("date_modified").toString());
+                                productImage.setDateModified(wooDateModified);
+                            } catch (ParseException e) {
+                                e.printStackTrace();
+                            }
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:399]: Separate Method

File: src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java (hunk 17 above)
Comment (Tejas-Choudekar) [reply src/main/java/com/example/resellerapp/resellerapp/services/WoocommerceRequestService.java:36]: this gives an error. Even if I comment line 32 or 35

File: bash.exe.stackdump [hunk 30]
```diff
@@ -0,0 +1,16 @@
+Stack trace:
```
Comment (arohablue) [inline bash.exe.stackdump:1]: why this file ?

File: build.gradle [hunk 31]
```diff
@@ -48,25 +50,33 @@ dependencies {
...
 	
-	
-	
-    
+	// https://projectlombok.org
+	compileOnly 'org.projectlombok:lombok:1.18.4'
+    annotationProcessor 'org.projectlombok:lombok:1.18.4'
```
Comment (arohablue) [inline build.gradle:78]: extra space is there before compileonly

File: src/main/java/com/example/resellerapp/resellerapp/helpers/ProductCategoryHelper.java [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+
+                if (wooParentId != 0) {
+
+                    if (categoryToSetAsParent != null && productCategory != null) {
+
+                        productCategory.setParentCategory(categoryToSetAsParent);
+                        productCategoryRepository.save(productCategory);
+                    }
+                }
+            }
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/helpers/ProductCategoryHelper.java:111]: move above logic to a new method.

File: src/main/java/com/example/resellerapp/resellerapp/interfaces/ProductPropertyHelperInterface.java [hunk 33]
```diff
@@ -0,0 +1,20 @@
+package com.example.resellerapp.resellerapp.interfaces;
+import java.util.List;
+import java.util.Map;
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/interfaces/ProductPropertyHelperInterface.java:3]: Do not need an interface for Helper. The interface is only used if we want to control classes or to promote code reusability between classes. For example, the interface will have methods that we want the class to override and implement the body. We should not use an interface if we expect only one class will extend it.

File: src/main/java/com/example/resellerapp/resellerapp/interfaces/ProductPropertyHelperInterface.java [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+    public Boolean layInAllProducts(List<Map<String, Object>> wooProducts)throws Exception;
+    public Boolean layInAllProductAttributes(List<Map<String, Object>> wooProducts);
+    public Boolean mapProductAndAttributes(List<Map<String, Object>> wooProducts);
+    public Boolean layInAllProductDimensions(List<Map<String, Object>> wooProducts);
+    public Boolean mapProductAndDimensions(List<Map<String, Object>> wooProducts);
+    public Boolean layInAllProductImages(List<Map<String, Object>> wooProducts);
+    public Boolean mapProductAndImages(List<Map<String, Object>> wooProducts);
+    public Boolean regularPriceSetter(String storeThisProductPrice, @Nullable String storeThisProductRegularPrice);
+    public int stockQuantitySetter(@Nullable Integer storeStockQuantity);
+    public String stockSetter(Boolean stockCheckerBoolean);
```
Comment (arohablue) [inline src/main/java/com/example/resellerapp/resellerapp/interfaces/ProductPropertyHelperInterface.java:17]: also, Just a Note:  All methods of the class to be declared in the interface makes no sense its the other way around. Only the methods we expect to be implemented by differently by different classes or methods we want the implementing class to have compulsory should be here

File: src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java (hunk 5 above)
Comment (arohablue) [reply src/main/java/com/example/resellerapp/resellerapp/models/ProductCategory.java:38]: update this, please.

**Pull request merged**